lib/fluent/plugin/out_secondary_file.rb

```diff
@@ -0,0 +1,151 @@
+#    distributed under the License is distributed on an "AS IS" BASIS,
+#    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    See the License for the specific language governing permissions and
+#    limitations under the License.
+#
+
+require "fileutils"
+require "fluent/plugin/file_util"
+require "fluent/plugin/output"
+require "fluent/config/error"
+
+module Fluent::Plugin
+  class SecondaryFileOutput < Output
+    Fluent::Plugin.register_output("secondary_file", self)
+
+    FILE_PERMISSION = 0644
+    DIR_PERMISSION = 0755
+
+    desc "The Path of the file."
+    config_param :path, :string
+    desc "The flushed chunk is appended to existence file or not."
+    config_param :append, :bool, default: false
+    config_param :compress, :enum, list: [:normal, :gz, :gzip], default: :normal
+
+    def configure(conf)
+      super
+
+      unless @as_secondary
+        raise Fluent::ConfigError, "This plugin can only be used in the <secondary> section"
+      end
+
+      configure_path!
+
+      test_path = generate_path(Time.now.strftime("%Y%m%d"))
+      unless Fluent::FileUtil.writable_p?(test_path)
+        raise Fluent::ConfigError, "out_secondary_file: `#{test_path}` is not writable"
+      end
+
+      @dir_perm = system_config.dir_permission || DIR_PERMISSION
+      @file_perm = system_config.file_permission || FILE_PERMISSION
+    end
+
+    def write(chunk)
+      id = generate_id(chunk)
+      path = generate_path(id)
+      FileUtils.mkdir_p File.dirname(path), mode: @dir_perm
+
+      case @compress
+      when :normal
+        File.open(path, "ab", @file_perm) {|f|
+          f.flock(File::LOCK_EX)
+          chunk.write_to(f)
+        }
+      when :gz, :gzip
+        File.open(path, "ab", @file_perm) {|f|
+          f.flock(File::LOCK_EX)
+          gz = Zlib::GzipWriter.new(f)
+          chunk.write_to(gz)
+          gz.close
+        }
+      end
+
+      path
+    end
+
+    private
+
+    def generate_id(chunk)
+      if chunk.metadata.empty?
+        chunk.chunk_id
+      else
+        extract_placeholders(@path, chunk.metadata)
+      end
+    end
+
+    def configure_path!
+      # Check @path includes ${tag} etc.
+      matched = @path.scan(/\${([\w.@-]+(\[\d+\])?)}/).flat_map(&:first) # to trim suffix [\d+]
+
+      if path_has_tags(matched) || path_has_time_format?
+        validate_path_is_comptible_with_primary_buffer!(matched)
+        @path_prefix = ""
+        @path_suffix = ".log"
+      else
+        if pos = @path.index('*')
+          @path_prefix = @path[0, pos]
+          @path_suffix = @path[pos+1..-1]
+        else
+          @path_prefix = @path + "."
+          @path_suffix = ".log"
+        end
+      end
+    end
+
+    def validate_path_is_comptible_with_primary_buffer!(matched)
+      raise "BUG: file path has imcompatible placeholder: Time" if !@chunk_key_time && path_has_time_format?
+      matched.each do |e|
```

Comment:
Is it needed to use loop & case? In configuration stage, small loops (inefficient code) doesn't matter in any way.
It looks enough to do:
* `@chunk_key_tag` xor `placeholders.any?{|placeholder| placeholder.matches(/tag pattern/) }`
* `@chunk_key_time` xor `has_time_format?(path)`
* `@chunk_keys` == rest placeholders (excluding tags)